### PR TITLE
improve error logging

### PR DESCRIPTION
**Pull request by leifwalsh**
* Refactors handle_request to stop using a callback
* Logs error messages a bit better

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -44,47 +44,63 @@
...
+    };
+    let request = match protocol::Request::parse(&buf[0..size_read]) {
+        Ok(x) => x,
+        Err(e) => {
+            debug!(log, "parsing request"; "err" => %e);
+            return;
+        }
+    };
+    let type_str = format!("{:?}", request.ty);
+    let log = log.new(o!("request_type" => type_str));
```
Comment (leifwalsh) [inline src/main.rs:82]: This is probably unnecessary but demonstrates logging context?

File: src/handlers.rs [hunk 2]
```diff
@@ -81,17 +76,15 @@ where
...
 }
 
 /// Send a user (passwd entry) back to the client, or a response indicating the
 /// lookup found no such user.
-fn send_user<F>(log: &Logger, user: Option<User>, mut send_slice: F) -> Result<()>
-where
-    F: FnMut(&[u8]) -> Result<()>,
-{
+fn serialize_user(log: &Logger, user: Option<User>) -> Result<Vec<u8>> {
     debug!(log, "got user"; "user" => ?user);
```
[the remark this answers is not in the document]
Comment (leifwalsh) [reply src/handlers.rs:86]: write a test for what?

File: src/main.rs [hunk 3]
```diff
@@ -44,47 +44,63 @@
...
+    let response = match handlers::handle_request(&log, &request) {
+        Ok(x) => x,
+        Err(e) => {
+            error!(log, "handling request"; "err" => %e);
```
[the remark this answers is not in the document]
Comment (leifwalsh) [reply src/main.rs]: it'll just show the error message from whatever error occurred

File: src/handlers.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (blinsay) [reply src/handlers.rs]: let's write some unit tests for `serialize_user` in another PR

File: src/handlers.rs [hunk 4]
```diff
@@ -33,35 +33,34 @@ use super::protocol::RequestType;
...
-{
+pub fn handle_request(log: &Logger, request: &protocol::Request) -> Result<Vec<u8>> {
     debug!(log, "handling request"; "request" => ?request);
     match request.ty {
         RequestType::GETPWBYUID => {
             let key = CStr::from_bytes_with_nul(request.key)?;
             let uid = atoi::<u32>(key.to_bytes()).context("invalid uid string")?;
             let user = User::from_uid(Uid::from_raw(uid))?;
-            send_user(log, user, send_slice)
+            debug!(log, "got user"; "user" => ?user);
```
Comment (blinsay) [inline src/handlers.rs:43]: for debug, might be useful to log which libc method we're using for the lookup

```rust
debug!(log, "User::from_uid"; "user" => ?user)
```

this is what @geofft wanted with `stringify!`. does slog have an equivalent of [`dbg`](https://doc.rust-lang.org/std/macro.dbg.html)?

Comment (leifwalsh) [reply src/handlers.rs:43]: I don't think it has a `dbg!` equivalent, but I did attach the method as context up in `handle_stream`

**Pull request merged**